### PR TITLE
Refactoring and cleanup around transaction appenders

### PR DESCRIPTION
**Pull request by chrisvest**
- Collapses the AbstractPhysicalTransactionAppender hierarchy.
- Removes non-batched appending and obsoletes the `batched_writes` configuration option.
- Rename a couple of methods for clarity.
- Move InMemoryLogChannel to test sources.

### CODE REVIEW

**Comment by tinwelint** [comment]
To start this off, I'd like to consider the convenience of putting this in 2.2... for future maintainability w/ regards to further changes made in 2.2 around this code. Rebasing on 2.2 should be a walk in the park.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionRepresentationCommitProcess.java [hunk 1]
```diff
@@ -62,59 +62,65 @@ public long commit( TransactionRepresentation transaction, LockGroup locks, Comm
...
+            return indexUpdatesValidator.validate( transaction, mode );
+        }
+        catch ( Throwable e )
+        {
+            throw new TransactionFailureException( ValidationFailed, e, "Validation of index updates failed" );
+        }
     }
 
-    private long commitTransaction( TransactionRepresentation tx, CommitEvent commitEvent ) throws TransactionFailureException
+    private long serialiseToLog(
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionRepresentationCommitProcess.java]: Feels unnecessary to introduce an additional term `serialise` when we have `append`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java [hunk 2]
```diff
@@ -129,12 +121,13 @@ public long append( TransactionRepresentation transaction, LogAppendEvent logApp
                 {
                     transactionId = transactionIdStore.nextCommittingTransactionId();
                     phase = 1;
-                    commit = append0( transaction, transactionId );
+                    commit = serializeToTransactionLog( transaction, transactionId );
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java]: Same here about `serialize` (which has a different spelling than above btw) --> `append` methinks.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java [hunk 3]
```diff
@@ -167,71 +160,186 @@ public Commitment append( TransactionRepresentation transaction, long expectedTr
                         "Received " + transaction + " with txId:" + expectedTransactionId +
                         " to be applied, but appending it ended up generating an unexpected txId:" + transactionId );
             }
-            return append0( transaction, transactionId );
+            return serializeToTransactionLog( transaction, transactionId );
         }
     }
 
-    private class TransactionCommitment implements Commitment
+    private static class TransactionCommitment implements Commitment
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java:167]: What would you say is the difference between having a non-static internal class and a static internal w/ reference to parent?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java [hunk 4]
```diff
@@ -167,71 +160,186 @@ public Commitment append( TransactionRepresentation transaction, long expectedTr
...
         }
+    }
 
-        @Override
-        public void transactionCommitted()
+    /**
+     * @return A TransactionCommitment instance with metadata about the committed transaction, such as whether or not this transaction
+     * contains any legacy index changes.
+     */
+    private TransactionCommitment serializeToTransactionLog(
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java]: Same here about `serialize` (which has a different spelling than above) --> `append` methinks.

**Comment by tinwelint** [comment]
Generally the changes look good and simplifies the code, good work and I agree about removing the non-batched version.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java (hunk 3 above)
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingTransactionAppender.java:167]: It's more obvious where the fields come from in the latter case. The two are less coupled, and this makes it easier to convert the inner class to a top level class, if we want to do that later.

**Comment by tinwelint** [comment]
Superseded by #4599 (for 2.2 instead)

**Pull request closed without merging**